speeduino/sensors.ino

```diff
@@ -707,3 +725,53 @@ uint16_t readAuxdigital(uint8_t digitalPin)
   tempReading = digitalRead(digitalPin); 
   return tempReading;
 } 
+
+/*
+ * Called by the 4 hz hardware timer, enables OPS+T interrupt
+ * when timer clicks, ISR will detach itself when it has got readings
+*/
+void readOPSt () {
+  attachInterrupt(digitalPinToInterrupt(pinOilSensorOPSt), oilSensorOPStISR, CHANGE);       
+}
+
+// ISR to decode PPM data from HELLA oil temperature and pressure sensor
+  static inline void oilSensorOPStISR() //Most ARM chips can simply call a function
+{
+  oilSensorOPStPulse.curEvent = micros();
+
+  if(oilSensorOPStPulse.lastLevel == 0 && digitalRead(pinOilSensorOPSt) ) // Last event was LOW and we have got a rising edge
+  {
+      oilSensorOPStPulse.offTime = oilSensorOPStPulse.curEvent - oilSensorOPStPulse.lastEvent;
+      oilSensorOPStPulse.totalTime = oilSensorOPStPulse.offTime + oilSensorOPStPulse.onTime;
+      oilSensorOPStPulse.lastLevel = 1;
+      oilSensorOPStPulse.lastEvent = oilSensorOPStPulse.curEvent;
+      if (oilSensorOPStPulse.totalTime <= 1024 && oilSensorOPStPulse.index == 0) 
+      {
+        oilSensorOPStPulse.index++;
+        oilSensorOPStPulse.gotSync = 1; 
+        oilSensorOPStData.status = (1024.0/oilSensorOPStPulse.totalTime)*oilSensorOPStPulse.onTime;
+      } 
+      else if (oilSensorOPStPulse.index == 1 && oilSensorOPStPulse.gotSync == 1) 
+      {
+        oilSensorOPStData.temperature = ((4096.0/oilSensorOPStPulse.totalTime)*oilSensorOPStPulse.onTime-128)/19.2-40;
```

Comment:
`oilSensorOPStData.temperature = fastMap(oilSensorOPStPulse.onTime, 128, 3968, 0, 160) - 40;`

Reply:
oilSensorOPStData.temperature = fastMap(oilSensorOPStPulse.onTime, 128, 3968, 0, 200); // 1-2 C less on 23C
oilSensorOPStData.temperature = fastMap((4096.0/oilSensorOPStPulse.totalTime)*oilSensorOPStPulse.onTime, 128, 3968, 0, 200); // Correct

I've tried various combinations using fastMap starting from the one you proposed with just partial success, the proportional correction of onTime based on the effective total pulse length totalTime compared to the ideal pulse length of 4096 is necessary and we don't really have a maximum value for it, as it can be less or more than the ideal (4096) and it depends on internal  sensor drifting, interfacing etc. Maybe you can help me find another way of optimization considering this.

Reply:
You need to get rid of those floats, this cannot be pulled in with those still there as the codebase does not allow their use.

If the second example you gave works, why not just shuffle things around to get rid of the floats? Eg

Replace:
`oilSensorOPStData.temperature = fastMap((4096.0/oilSensorOPStPulse.totalTime)*oilSensorOPStPulse.onTime, 128, 3968, 0, 200);`

With:
`oilSensorOPStData.temperature = fastMap((4096UL*oilSensorOPStPulse.onTime)/oilSensorOPStPulse.totalTime, 128, 3968, 0, 200);`

Reply:
On the datasheet show a minimum width of 128 and a maximum 3968, if only the high level width matters, why you have to divide 4096 by total time?

Reply:
![IMG_20201229_072756](https://user-images.githubusercontent.com/22614869/103277282-6fdc3280-49a7-11eb-925c-4ca7dda5eaa4.jpg)
This graph shows a period of 4096uS

![IMG_20201229_072736](https://user-images.githubusercontent.com/22614869/103277372-9e5a0d80-49a7-11eb-8d51-732c2b2e8fb9.jpg)

And those clearly shows the minimum and maximum width of a valid signal, as you can see, minimum is 128 for temperature and pressure, same for maximum of 3968

Reply:
If the total time is 4096 don't make any sense use this on math, `4096/totalTime` will be `4096/4096` and it can be ignored, another thing, only assume new values if S1 is smaller than 384uS

Reply:
> On the datasheet show a minimum width of 128 and a maximum 3968, if only the high level width matters, why you have to divide 4096 by total time?

4096 is the theoretical unit of time (theoretical totalTime) into where and in relation to which the symbol (theoretical onTime) is meaningful. 
In reality both drift (of a proportional amount), by various reasons, temperature, voltage, communication, interrupt firing time, computational time and so on. 
So, to be aware of the amount of this drift we need to compare a measured value with its fixed theoretical counterpart: totalTime and 4096, their difference will tell us the amount of drift we are having and considering that we can understand how our real onTime was affected by this drift.

We cannot directly (succesfully) use onTime because its only meaningful related to the unit of time (4096) and the unit of time is 4096 only in theory.

I've tried Josh suggested resolution compared to the original one and I saw no differences, so I'm going to implement it that way for all the three pulses.

I'd like to implement sensor fault codes on TS and oil temperature based functions but maybe better to close this one first and do that in a future pull request.

Thank you for your help!